File: python/0079_Word_Search.py
class Solution:
    def exist(self, board: List[List[str]], word: str) -> bool:
        ROWS, COLS = len(board), len(board[0])
        path = set()

        def dfs(r, c, i):

            if i == len(word):
                return True
            
            if r < 0 or r >= ROWS or c < 0 or c >= COLS or (r, c) in path or word[i] != board[r][c]:
                return False
            
            path.add((r, c))
            
            for dx, dy in [(0, 1), (1, 0), (-1, 0), (0, -1)]:

                if dfs(r + dx, c + dy, i + 1):
                    return True
            path.remove((r, c))



        for r in range(ROWS):

            for c in range(COLS):

                if dfs(r, c, 0):
                    return True
        
        return False

File: java/0079_Word_Search.java
class Solution {
    int ROWS, COLS;
    boolean[][] path;

    public boolean exist(char[][] board, String word) {
        ROWS = board.length;
        COLS = board[0].length;
        path = new boolean[ROWS][COLS];

        for (int r = 0; r < ROWS; r++) {
            for (int c = 0; c < COLS; c++) {
                
                if (dfs(board, word, r, c, 0)) return true;
            }
        }

        return false;
    }

    private boolean dfs(char[][] board, String word, int r, int c, int i) {
        int[][] directions = {{0, 1}, {1, 0}, {0, -1}, {-1, 0}};
        if (i == word.length()) return true;

        if (r < 0 || r >= ROWS || c < 0 || c >= COLS || path[r][c] || board[r][c] != word.charAt(i)) return false;

        path[r][c] = true;

        for (int[] dir : directions) {
            int dx = dir[0], dy = dir[1];

            if (dfs(board, word, r + dx, c + dy, i + 1)) {
                return true;
            }
        }

        path[r][c] = false;
        return false;
    }
}